numpy/core/shape_base.py

```diff
@@ -179,6 +170,92 @@ def atleast_3d(*arys):
         return res
 
 
+def atleast_nd(ary, n, pos='left'):
+    """
+    View input as array with at least `n` dimensions.
+
+    Parameters
+    ----------
+    ary : array_like
+        The input array. Non-array inputs are converted to arrays.
+        Arrays that already have `n` or more dimensions are preserved.
+    n : scalar
+        The minimum number of dimensions required.
+    pos : {'left', 'right'}, optional
+        Where to add new dimensions. All new dimensions are added on the
+        same side of the shape. Default is 'left'.
+
+    Returns
+    -------
+    res : ndarray
+        An array with ``res.ndim >= n``. A view is returned for array
+        inputs. Dimensions are prepended if `pos` is 'left', so for
+        example, a 1-D array of shape ``(N,)`` with ``n=4`` becomes a
+        view of shape ``(1, 1, 1, N)``. Dimensions are appended if
+        `pos` is 'right', so for example a 2-D array of shape ``(M, N)``
+        becomes a view of shape ``(M, N, 1, 1)`` when ``n=4``.
+
+    See Also
+    --------
+    atleast_1d, atleast_2d, atleast_3d
+
+    Notes
+    -----
+    This function does not follow the convention of the other atleast_*d
+    functions in numpy in that it only accepts a single array argument.
+    To process multiple arrays, use a comprehension or loop around the
+    function call. See examples below.
+
+    Setting ``pos='left'`` is equivalent to how the array would be
+    interpreted by numpy's broadcasting rules. There is no need to call
+    this function for simple broadcasting. This is also roughly
+    equivalent to ``np.array(ary, copy=False, subok=True, ndmin=n)``.
+
+    It is easy to create functions for specific dimensions similar to
+    the other atleast_*d functions using Python's `functools.partial`
+    function. An example is shown below.
+
+    Examples
+    --------
+    >>> np.atleast_nd(3.0, 4)
+    array([[[[ 3.]]]])
+
+    >>> x = np.arange(3.0)
+    >>> np.atleast_nd(x, 2).shape
+    (1, 3)
+
+    >>> x = np.arange(12.0).reshape(4, 3)
+    >>> np.atleast_nd(x, 5).shape
+    (1, 1, 1, 4, 3)
+    >>> np.atleast_nd(x, 5).base is x.base
+    True
+
+    >>> [np.atleast_nd(x) for x in ((1, 2), [[1, 2]], [[[1, 2]]])]:
+    [array([[1, 2]]), array([[1, 2]]), array([[[1, 2]]])]
+
+    >>> np.atleast_nd((1, 2), 5, 'start').shape
+    (1, 1, 1, 1, 2)
+    >>> np.atleast_nd((1, 2), 5, 'end').shape
+    (2, 1, 1, 1, 1)
+
+    >>> from functools import partial
+    >>> atleast_4d = partial(np.atleast_nd, n=4)
+    >>> atleast_4d([1, 2, 3])
+    [[[[1, 2, 3]]]]
+    """
+    if pos == 'left':
+        res = array(ary, copy=False, subok=True, ndmin=n)
```

Comment:
This fails on duck arrays - I'd be inclined to just support anything with a `.ndim` property

Reply:
(and go back to the slicing for either case)

Reply:
But we call `array(ary, copy=False, subok=True)` in every case anyway. I'm not sure I see what the difference would be. I'm fine with going back to slicing for both cases, especially with `pos` about to be an integer now.